controllers/authControllerLoginController.test.js

```diff
@@ -0,0 +1,223 @@
+        };
+    });
+
+   
+    describe('should return error messages for its input validations if', () => {
+        it('email is empty', async () => {
+            //ARRANGE
+            const req = {
+                body: {
+                    email: "",
+                    password: "",
+                }
+            };
+    
+            //ACTION
+            await loginController(req, res);
+
+            //ASSERT
+            expect(res.status).toHaveBeenCalledTimes(1);
+            expect(res.status).toHaveBeenCalledWith(404);
+            expect(res.send).toHaveBeenCalledTimes(1);
+            expect(res.send).toHaveBeenCalledWith({ success: false, message: "Invalid email or password" });
+        });
+
+
+        it('password is empty', async () => {
+            //ARRANGE
+            const req = {
+                body: {
+                    email: "james@gmail.com",
+                    password: "",
+                }
+            };
+    
+            //ACTION
+            await loginController(req, res);
+
+            //ASSERT
+            expect(res.status).toHaveBeenCalledTimes(1);
+            expect(res.status).toHaveBeenCalledWith(404);
+            expect(res.send).toHaveBeenCalledTimes(1);
+            expect(res.send).toHaveBeenCalledWith({ success: false, message: "Invalid email or password" });
+        });
+    });
+
+
+    describe('should not allow user to log in if', () => {
+        let req;
+        beforeEach(() => {
+            req = {
+                body: {
+                    email: "james@gmail.com",
+                    password: "james123",
+                }
+            };
+        });
+
+
+        //NEVER PASS
+        it('email is not registered in database', async () => {
+            //ARRANGE
+            userModel.findOne.mockImplementation((queryInput) => {
+                return Promise.resolve(null); //Email is not found
+            });
+
+            //ACTION
+            await loginController(req, res);
+
+            //ASSERT
+            expect(res.status).toHaveBeenCalledTimes(1);
+            expect(res.status).toHaveBeenCalledWith(404);
+            expect(res.send).toHaveBeenCalledTimes(1);
+            expect(res.send).toHaveBeenCalledWith({ success: false, message: "Email is not registered" });
+        });
+
+        
+        //NEVER PASS
+        it('password is incorrect', async () => {
+            //ARRANGE
+            userModel.findOne.mockImplementation((queryInput) => {
+                return Promise.resolve({}); //Email is found, not null
+            });
+            authHelper.comparePassword.mockImplementation((passwordToCheck, correctPassword) => {
+                return Promise.resolve(false); //Password is incorrect
+            });
+
+            //ACTION
+            await loginController(req, res);
+
+            //ASSERT
+            expect(res.status).toHaveBeenCalledTimes(1);
+            expect(res.status).toHaveBeenCalledWith(404);
```

Comment:
I think the code should be 401, which represents unauthorised. But not too sure though.

Reply:
Yeah, agree with you that `401` might be more appropriate. I originally left it as 404 because I thought the developer might have done that on purpose. However, I have changed to `401` accordingly, will just state it as a potential bug in the write-up.